src/main/java/com/kalessil/phpStorm/phpInspectionsEA/inspectors/forEach/ForeachSourceInspector.java

```diff
@@ -54,33 +73,35 @@ public String getShortName() {
     public PsiElementVisitor buildVisitor(@NotNull final ProblemsHolder holder, final boolean isOnTheFly) {
         return new BasePhpElementVisitor() {
             @Override
-            public void visitPhpForeach(@NotNull ForeachStatement foreach) {
-                final PsiElement source = ExpressionSemanticUtil.getExpressionTroughParenthesis(foreach.getArray());
-                if (source instanceof PhpTypedElement && !isEnsuredByPyParentIf(foreach, source)) {
-                    this.analyseContainer(source);
+            public void visitPhpForeach(@NotNull final ForeachStatement foreachStatement) {
+                final PsiElement source = ExpressionSemanticUtil.getExpressionTroughParenthesis(foreachStatement.getArray());
+
+                if ((source instanceof PhpTypedElement) && !isEnsuredByPyParentIf(foreachStatement, source)) {
+                    analyseContainer(source);
                 }
             }
 
             /* should cover is_array/is_iterable in direct parent if of the loop, while PS types resolving gets improved */
-            private boolean isEnsuredByPyParentIf(@NotNull ForeachStatement foreach, @NotNull PsiElement source) {
+            private boolean isEnsuredByPyParentIf(@NotNull final PhpPsiElement foreach, @NotNull final PsiElement source) {
                 boolean result = false;
+
                 if (foreach.getPrevPsiSibling() == null) {
-                    final PsiElement ifCandidate = foreach.getParent() instanceof GroupStatement ? foreach.getParent().getParent() : null;
-                    final PsiElement conditions;
-                    if (ifCandidate instanceof If) {
-                        conditions = ((If) ifCandidate).getCondition();
-                    } else if (ifCandidate instanceof ElseIf) {
-                        conditions = ((ElseIf) ifCandidate).getCondition();
-                    } else {
-                        conditions = null;
+                    final PsiElement ifCandidate = (foreach.getParent() instanceof GroupStatement) ? foreach.getParent().getParent() : null;
+                    PsiElement       conditions  = null;
+
+                    if (ifCandidate instanceof ControlStatement) {
```

Comment:
Catches lots of other constructs (not only If, ElseIf) - is it intended?

Reply:
It will catch some additional constructs, really. But will support things like that:

```php
switch (true){
    case is_array($value):
        foreach ($value as $item) {
        }
        break;
}
```

And too to `while()`, `do while()` and `for()` conditionals. But we can limit to works strictly with that types, once that it will includes `catch()` and `finally()` that don't have really a conditional.

Reply:
Let's stick to If, ElseIf - I'd like to minimize this inspection modification - it's very problematic one =_